eip-0031.md

```diff
@@ -0,0 +1,154 @@
+                val babelTokensDifference: Long = recreatedBabelBox.tokens(0)._2 - babelTokensBefore
+
+                allOf(Coll(
+                    (babelTokensDifference * ergPricePerToken >= nanoErgsDifference),
+                    (nanoErgsDifference >= 0)
+                ))
+
+            }
+
+            allOf(Coll(
+                validBabelFeeBoxRecreation,
+                validBabelFeeExchange
+            ))
+
+        }
+
+        sigmaProp(validBabelFeeSwap)
+
+    } else {
+
+        // ===== Perform Babel Fee Box Withdrawl ===== //
+        babelFeeBoxCreator
+
+    }
+
+}
+```
+Compilation of babel fee box smartcontracts for different tokens of interest will result in different P2S addresses, leading to each token having a unique corresponding P2S babel fee box smartcontract address.
+
+### Output position
+
+The position of the recreated Babel Box in transaction's outputs must be set as a context extension variable in the spending Babel Box.
+
+1. Context variable `0`
+    * Type: `SInt`
+    * Value: position of the recreated Babel Box in transaction's outputs
+
+### Parameters
+Parameters (creator’s public key and price) are specified via registers, meaning the resulting babel fee boxes from different creators will always belong to the same P2S address, which will improve their searchability.
+
+#### Registers:
+1. Register R4
+    * Type: `SigmaProp`
+    * Value: creator's public key
+2. Register R5
+    * Type: `SLong`
+    * Value: how much nanoErgs is the creator willing to pay for one babel token
+
+Babel fee box creator is able to spent the babel box in any circumstances.
+
+Other users on the other hand can spend this box as input to their transaction only when they also recreate it as output of their transaction with the very same registers (R4, R5) values together with insertion of a required amount of babel tokens (the amount of inserted babel tokens multiplied by the price specified in the R5 register has to be equal to or bigger than the amount of nanoErgs spent from the babel fee box).
+
+Wallets implementation
+--------------------------------------------
+
+Wallet developers will need to decide whether they want to support EIP-0031 or not. If they do decide to support this standard, they should also decide on which tokens they want to support (this could be done based on user requirements – e.g. implementing big stablecoins or “meme” coins, etc.), as this could be more convenient than supporting all tokens.
+
+As P2S addresses belonging to specific token of interest stay the same, these addresses could be easily “hardcoded” when supporting only a few tokens. If the developers decide to support any token, the previously mentioned smartcontract for each token which user holds should be compiled and the availability of babel fee boxes (UTXOs) for the specific tokens of interest in the blockchain should be subsequently checked.
```

Comment:
As pointed out before, the P2S addresses can change when a new compiler version is used. The addresses for a token should be defined in the EIP to avoid having diverging addresses in the future.

Reply:
I see no need for that.

- Address will change from token to token;
- It's very unlikely to have compiler serializations changes, I could be wrong but it requires a hard fork to happen.

Reply:
I think there is some misconception, in the EIP text "smartcontract for each token which user holds should be compiled". 
When we talk about smart contracts we should really mean contract templates (see [EIP-5](https://github.com/ergoplatform/eips/blob/master/eip-0005.md)) or at least ErgoTree. 

ErgoScript should be out of question as well as any `compilation`. Compilation should not be part of any on-chain protocol. First, it is unstable, second it is not standardized.

Contract Template capture the parameterized nature of smart contracts and ideally any App which want to support pre-defined list of contracts should actually use a predefined list of Contract Templates.

Potentially unlimited P2S addresses can belong to the same Contract Template. 
At the same time, any protocol usually require fixed list of Contract Tempates.

Reply:
Oh, I didn't know about EIP-5, contract template sounds like a great addition to have on EIP text. Thanks for the clarification )

Reply:
> I see no need for that.

I see a high need for a more defined approach, as the contract in the EIP does not compile for me. So changes are needed on my side to get it working, and the chances are high that I end up with another ergotree than others.
Maybe EIP-5 is a solution as well.

Reply:
What changes were needed?

Reply:
Contract fixes pushed

Reply:
Yup

Reply:
"Contract Template" subsection added